### PR TITLE
Manually merge 2.10.x into 2.11.x

### PR DESCRIPTION
**Pull request by greg0ire**
Review carefully, I'm not sure I didn't mess up

### CODE REVIEW

**Comment by morozov** [comment]
This looks a bit different from my result (https://github.com/morozov/dbal/commit/ddf56413f254968ce14cac5161b6ab3bb4a42c96).

The changes in AbstractSQLiteDriver.php are missing but they exist in #4255. You can go over is mismatch individually or just use my version.

```diff
diff --git a/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php b/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php
index a8bbf60ce..e14a2a76c 100644
--- a/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php
+++ b/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php
@@ -7,7 +7,6 @@ use Doctrine\DBAL\Driver;
 use Doctrine\DBAL\Driver\DriverException as DeprecatedDriverException;
 use Doctrine\DBAL\Exception\ConnectionException;
 use Doctrine\DBAL\Exception\DriverException;
-use Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException;
 use Doctrine\DBAL\Exception\InvalidFieldNameException;
 use Doctrine\DBAL\Exception\LockWaitTimeoutException;
 use Doctrine\DBAL\Exception\NonUniqueFieldNameException;
@@ -84,10 +83,6 @@ abstract class AbstractSQLiteDriver implements Driver, ExceptionConverterDriver
             return new ConnectionException($message, $exception);
         }
 
-        if (strpos($exception->getMessage(), 'FOREIGN KEY constraint failed') !== false) {
-            return new ForeignKeyConstraintViolationException($message, $exception);
-        }
-
         return new DriverException($message, $exception);
     }
 
diff --git a/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php b/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php
index 65d4c12de..78a3ac554 100644
--- a/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php
+++ b/tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php
@@ -3,7 +3,6 @@
 namespace Doctrine\Tests\DBAL\Platforms;
 
 use Doctrine\Common\EventManager;
-use Doctrine\DBAL\DBALException;
 use Doctrine\DBAL\Events;
 use Doctrine\DBAL\Exception;
 use Doctrine\DBAL\Platforms\AbstractPlatform;
diff --git a/tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php b/tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php
index c40b4a53f..7a0983170 100644
--- a/tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php
+++ b/tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php
@@ -76,7 +76,7 @@ class CreateSchemaSqlCollectorTest extends TestCase
 
     public function testAcceptsForeignKeyDoesNotSupportCreateDropForeignKeyConstraints(): void
     {
-        $this->platformMock->method('supportsForeignKeyConstraints')
+        $this->platformMock->method('supportsCreateDropForeignKeyConstraints')
             ->willReturn(false);
 
         $table      = $this->createTableMock();
@@ -87,6 +87,19 @@ class CreateSchemaSqlCollectorTest extends TestCase
         self::assertEmpty($this->visitor->getQueries());
     }
 
+    public function testAcceptsForeignKeySupportsCreateDropForeignKeyConstraints(): void
+    {
+        $this->platformMock->method('supportsCreateDropForeignKeyConstraints')
+            ->willReturn(true);
+
+        $table      = $this->createTableMock();
+        $foreignKey = $this->createForeignKeyConstraintMock();
+
+        $this->visitor->acceptForeignKey($table, $foreignKey);
+
+        self::assertSame(['foo'], $this->visitor->getQueries());
+    }
+
     public function testAcceptsSequences(): void
     {
         $sequence = $this->createSequenceMock();
```

File: tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php [hunk 1]
```diff
@@ -87,19 +87,6 @@ public function testAcceptsForeignKeyDoesNotSupportCreateDropForeignKeyConstrain
         self::assertEmpty($this->visitor->getQueries());
     }
 
-    public function testAcceptsForeignKeySupportsCreateDropForeignKeyConstraints(): void
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php:90]: This chunk should be reverted.

**Review by morozov — CHANGES_REQUESTED** [review]

File: tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php [hunk 2]
```diff
@@ -261,6 +258,18 @@ public function testGeneratesConstraintCreationSql(): void
...
     }
 
+    public function testGeneratesForeignKeySqlOnlyWhenSupportingForeignKeys(): void
+    {
+        $fk = new ForeignKeyConstraint(['fk_name'], 'foreign', ['id'], 'constraint_fk');
+
+        if ($this->platform->supportsForeignKeyConstraints()) {
+            self::assertIsString($this->platform->getCreateForeignKeySQL($fk, 'test'));
+        } else {
+            $this->expectException(DBALException::class);
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:268]: This is deprecated. It should expect `Exception`.

File: tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php [hunk 3]
```diff
@@ -76,7 +76,7 @@ public function testAcceptsTable(): void
 
     public function testAcceptsForeignKeyDoesNotSupportCreateDropForeignKeyConstraints(): void
     {
-        $this->platformMock->method('supportsCreateDropForeignKeyConstraints')
+        $this->platformMock->method('supportsForeignKeyConstraints')
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/Visitor/CreateSchemaSqlCollectorTest.php:79]: This also warrants the test name change. Should have been done in `2.10.x` but let's at least fix it here.

**Comment by greg0ire** [comment]
I run `vendor/bin/phpcbf on top of morozov@d26a383, it should be OK, thanks for the review!

**Review by morozov — APPROVED** [review]

**Pull request merged**